streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java

```diff
@@ -160,12 +160,17 @@ public StateStore getStateStore(final String name) {
                                final V value,
                                final To to) {
         final ProcessorNode previousNode = currentNode();
-        final long currentTimestamp = recordContext.timestamp();
+        final ProcessorRecordContext previousContext = recordContext;
 
         try {
             toInternal.update(to);
             if (toInternal.hasTimestamp()) {
-                recordContext.setTimestamp(toInternal.timestamp());
+                recordContext = new ProcessorRecordContext(
+                    toInternal.timestamp(),
+                    recordContext.offset(),
+                    recordContext.partition(),
+                    recordContext.topic(),
+                    recordContext.headers());
```

Comment:
@vvcephei I put this fix to make `ProcessorRecordContext` immutable -- however, after you comment about headers being mutable, I am wondering if this is an issue we need to address here or not. Maybe not, because it's a general issue and the users responsibility to deep-copy headers is they are modified. Just wanted to double check and point it out.